File: 2-add-two-numbers/add-two-numbers.cpp
/**
 * Definition for singly-linked list.
 * struct ListNode {
 *     int val;
 *     ListNode *next;
 *     ListNode() : val(0), next(nullptr) {}
 *     ListNode(int x) : val(x), next(nullptr) {}
 *     ListNode(int x, ListNode *next) : val(x), next(next) {}
 * };
 */
class Solution {
public:
    ListNode* addTwoNumbers(ListNode* l1, ListNode* l2) {
        ListNode* t1 = l1;
        ListNode* t2 = l2;
        ListNode* dummy = new ListNode(0);
        ListNode* current = dummy;
        int carry = 0;
        while (t1 != nullptr || t2 != nullptr) {
            int sum = carry;
            if (t1 != nullptr) {
                sum += t1->val;
                t1 = t1->next;
            }
            if (t2 != nullptr) {
                sum += t2->val;
                t2 = t2->next;
            }
            ListNode* newNode = new ListNode(sum % 10);
            carry = sum / 10;
            current->next = newNode;
            current = current->next;
        }
        if (carry > 0) {
            ListNode* newNode = new ListNode(carry);
            current->next = newNode;
        }
        return dummy->next;
    }
};

File: 2-add-two-numbers/add-two-numbers.js
/**
 * Definition for singly-linked list.
 * function ListNode(val, next) {
 *     this.val = (val===undefined ? 0 : val)
 *     this.next = (next===undefined ? null : next)
 * }
 */
/**
 * @param {ListNode} l1
 * @param {ListNode} l2
 * @return {ListNode}
 */
var addTwoNumbers = function (l1, l2) {
    let temp1 = l1, temp2 = l2;
    let dummyNode = new ListNode(0);
    let current = dummyNode;
    let carry = 0;
    while (temp1 !== null || temp2 !== null) {
        let sum = carry;
        if (temp1) {
            sum += temp1.val;
            temp1 = temp1.next;
        }
        if (temp2) {
            sum += temp2.val;
            temp2 = temp2.next;
        }
        let newNode = new ListNode(sum % 10);
        carry = Math.floor(sum / 10);
        current.next = newNode;
        current = current.next;
    }
    if (carry > 0) {
        let newNode = new ListNode(carry);
        current.next = newNode;
    }
    return dummyNode.next;
};